File: BSvsZP-Common -Java/Messages/PlayingFieldReply.java
package Messages;

import java.io.NotActiveException;
import java.net.UnknownHostException;

import org.omg.CORBA.portable.ApplicationException;

import Common.ByteList;
import Common.PlayingFieldLayout;

public class PlayingFieldReply extends Reply {

    public PlayingFieldLayout Layout;
    private static int MinimumEncodingLength;

    protected PlayingFieldReply() {
    }

    public PlayingFieldReply(PossibleStatus status, PlayingFieldLayout layout, String... note) {
        super(Reply.PossibleTypes.PlayingFieldReply, status, (note.length == 1 ? note[0] : null));
        Layout = layout;
    }

    public static PlayingFieldReply Create(ByteList messageBytes) throws Exception {
        PlayingFieldReply result = null;

        if (messageBytes == null || messageBytes.getRemainingToRead() < getMinimumEncodingLength()) {
            throw new ApplicationException("Invalid message byte array", null);
        }
        if (messageBytes.PeekInt16() != MESSAGE_CLASS_IDS.PlayingFieldReply.getValue()) {
            throw new ApplicationException("Invalid message class id", null);
        } else {
            result = new PlayingFieldReply();
            result.Decode(messageBytes);
        }

        return result;
    }

    @Override
    public void Encode(ByteList bytes) throws NotActiveException, UnknownHostException, Exception {
        bytes.Add((short) MESSAGE_CLASS_IDS.PlayingFieldReply.getValue());                           // Write out this class id first
        bytes.update();
        short lengthPos = bytes.getCurrentWritePosition();   // Get the current write position, so we
        // can write the length here later
        bytes.Add((short) 0);                           // Write out a place holder for the length
        bytes.update();
        super.Encode(bytes);                             // Encode stuff from base class
        bytes.update();
        bytes.Add(Layout);
        bytes.update();
        short length = (short) (bytes.getCurrentWritePosition() - lengthPos - 2);
        bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        

    }

    @Override
    public void Decode(ByteList bytes) throws Exception {
        short objType = bytes.GetInt16();
        bytes.update();
        short objLength = bytes.GetInt16();
        bytes.update();

        bytes.SetNewReadLimit(objLength);
        bytes.update();

        super.Decode(bytes);
        bytes.update();
        Layout = (PlayingFieldLayout) bytes.GetDistributableObject();
        bytes.update();
        bytes.RestorePreviosReadLimit();
        bytes.update();
    }

    public PlayingFieldLayout getLayout() {
        return Layout;
    }

    public void setLayout(PlayingFieldLayout layout) {
        Layout = layout;
    }

    @Override
    public short getClassId() {
        return (short) MESSAGE_CLASS_IDS.PlayingFieldReply.getValue();
    }

    public static int getMinimumEncodingLength() {
        MinimumEncodingLength = 4 // Object header
                + PlayingFieldLayout.getMinimumEncodingLength();
        return MinimumEncodingLength;
    }

    @Override
    public int compareTo(Object o) {
        // TODO Auto-generated method stub
        return 0;
    }

    @Override
    public MESSAGE_CLASS_IDS MessageTypeId() {
        return Message.MESSAGE_CLASS_IDS.PlayingFieldReply;
    }

}

File: BSvsZP-Common/Messages/PlayingFieldReply.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Common;

namespace Messages
{
    public class PlayingFieldReply : Reply
    {
        #region Private Properties
        private static Int16 ClassId { get { return (Int16)MESSAGE_CLASS_IDS.PlayingFieldReply; } }
        #endregion

        #region Public Properties
        public override Message.MESSAGE_CLASS_IDS MessageTypeId() { return (Message.MESSAGE_CLASS_IDS)ClassId; }

        public PlayingFieldLayout Layout { get; set; }
        public static new int MinimumEncodingLength
        {
            get
            {
                return 4                // Object header
                       + PlayingFieldLayout.MinimumEncodingLength;
            }
        }
        #endregion
        
        #region Constructors and Factory Methods
        /// <summary>
        /// Default message constructor, used by Factory methods (i.e., the Create methods)
        /// </summary>

        protected PlayingFieldReply() { }

        /// <summary>
        /// Constructor used by all specializations, which in turn are used by
        /// senders of a message 
        /// </summary>
        /// <param name="conversationId">conversation id</param>
        /// <param name="status">Status of the ack/nak</status>
        /// <param name="note">error message or note</note>       
        public PlayingFieldReply(PossibleStatus status, PlayingFieldLayout layout, string note = null) :
            base(Reply.PossibleTypes.PlayingFieldReply, status, note)
        {
            Layout = layout;
        }

        /// <summary>
        /// Factor method to create a message from a byte list
        /// </summary>
        /// <param name="messageBytes">A byte list from which the message will be decoded</param>
        /// <returns>A new message of the right specialization</returns>
        new public static PlayingFieldReply Create(ByteList messageBytes)
        {
            PlayingFieldReply result = null;

            if (messageBytes==null || messageBytes.RemainingToRead<MinimumEncodingLength)
                throw new ApplicationException("Invalid message byte array");
            if (messageBytes.PeekInt16() != ClassId)
                throw new ApplicationException("Invalid message class id");
            else
            {
                result = new PlayingFieldReply();
                result.Decode(messageBytes);
            }

            return result;
        }
        #endregion

        #region Encoding and Decoding methods

        override public void Encode(ByteList bytes)
        {
            bytes.Add(ClassId);                           // Write out this class id first

            Int16 lengthPos = bytes.CurrentWritePosition;   // Get the current write position, so we
                                                                // can write the length here later
            bytes.Add((Int16) 0);                           // Write out a place holder for the length

            base.Encode(bytes);                             // Encode stuff from base class

            bytes.Add(Layout);

            Int16 length = Convert.ToInt16(bytes.CurrentWritePosition - lengthPos - 2);
            bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        

        }

        override public void Decode(ByteList bytes)
        {
            Int16 objType = bytes.GetInt16();
            Int16 objLength = bytes.GetInt16();

            bytes.SetNewReadLimit(objLength);

            base.Decode(bytes);

            Layout = bytes.GetDistributableObject() as PlayingFieldLayout;
            
            bytes.RestorePreviosReadLimit();
        }

        #endregion

    }
}
